File: 파이썬 알고리즘 인터뷰/7장 선형 자료구조/두 수의 합.py
class Solution:
    # 나의 풀이 (Runtime: 36ms, Memory: 14.5MB)
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        hash_map = {}
        
        # 해시 테이블을 구성한다.
        for idx, num in enumerate(nums):
            hash_map[num] = idx
            
        for idx, num in enumerate(nums):
            num_to_find = target - num
            
            # 목표 값에서 첫번째 값을 뺀 결과 값을 Key로 사용, 중복 사용 제외.
            if num_to_find in hash_map and idx != hash_map[num_to_find]:
                return [idx, hash_map[num_to_find]]

File: 파이썬 알고리즘 인터뷰/7장 선형 자료구조/두 수의 합.js
// 브루트포스를 이용한 풀이 (Runtime: 72ms, Memory: 38.6MB)
const twoSumBruteForce = function(nums, target) {
  for (let i = 0; i < nums.length; i++) {
    for (let j = i + 1; j < nums.length; j++) {
      if (nums[i] + nums[j] === target) {
        return [i, j];
      }
    }
  } 
};

// Object 기반의 Hash를 이용한 풀이 (Runtime: 84ms, Memory: 41.3MB)
 const twoSumHashMap = function(nums, target) {
  // reduce()를 이용해 해시테이블을 생성한다.
  const hashMap = nums.reduce((hash, num, idx) => {
    hash[num] = idx;
    return hash;
  }, {});

  for (let idx = 0; idx < nums.length; idx++) {
    const numToFind = target - nums[idx];
    
    // 목표 값에서 첫번째 값을 뺀 값을 Key로 이용해 Hash를 조회한다.
    if (hashMap.hasOwnProperty(numToFind) && idx !== hashMap[numToFind]) {
      return [idx, hashMap[numToFind]];
    }
  }
};